src/Workspaces/Core/Portable/Utilities/AsyncLazy`1.cs

```diff
@@ -508,13 +505,30 @@ private void OnAsynchronousRequestCancelled(object state)
             }
         }
 
-        // Using inheritance instead of wrapping a TaskCompletionSource to avoid a second allocation
-        private class Request : TaskCompletionSource<T>
+        private sealed class Request
         {
+            /// <summary>
+            /// The <see cref="CancellationToken"/> associated with this request. This field will be initialized before
+            /// any cancellation is observed from the token.
+            /// </summary>
+            private CancellationToken _cancellationToken;
             private CancellationTokenRegistration _cancellationTokenRegistration;
 
+            // We use a AsyncTaskMethodBuilder so we have the ability to cancel the task with a given cancellation token
+            // TODO: remove this once we're on .NET 4.6 and can move back to using TaskCompletionSource
+            private AsyncTaskMethodBuilder<T> _taskBuilder;
```

Comment:
Consider adding a comment that this should never be made readonly; as it's a mutable struct, and as its task is lazily-initialized, making it readonly would lead to it never being initialized.